### PR TITLE
new apiserver modelgeneration: add, switch, advance

### PR DESCRIPTION
**Pull request by hmlanigan**
## Description of change

- New apiserver for modelgeneration, includes AddGeneration, AdvanceGeneration, and SwitchGeneration, and related unit tests.
- Unit tests for api/modelgeneration added for  AddGeneration, AdvanceGeneration, and SwitchGeneration.
- Updated cmd and api for AddGeneration, AdvanceGeneration, and SwitchGeneration to take model tags.
- Moved MockResources and MockAuthorizer to api/base for multiple package use.
- Added MockAPICallerCloser.
- Related changes for CancelGeneration will be forth coming.

## QA steps

1. export JUJU_DEV_FEATURE_FLAGS=generations
2. juju bootstrap localhost
2052.  juju deploy percona-cluster mysql
2056.  juju add-generation
    target generation set to next
2057.  juju switch-generation current
    changes dropped and target generation set to current
2059.  juju advance mysql
    ERROR next generation is not active
2060.  juju switch-generation next
    target generation set to next
2061.  juju advance mysql
2062.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
2063.  juju add-unit mysql
2064.  juju add-generation
    target generation set to next
2065.  juju advance mysql/1
2068.  juju advance mysql/0
2069.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
2070.  juju add-generation
    target generation set to next
071.  juju advance mysql/0
2072.  juju advance mysql
2073.  juju switch-generation next
    ERROR cannot switch to next generation, as none exists
4. run `juju-db.bash`
2. db.generations.find(), should report 3 docs with: active = false, completed with an non zero NumberLong.

## Documentation changes

In the future

### CODE REVIEW

File: api/modelgeneration/modelgeneration_test.go [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+func (s *modelGenerationSuite) TestAddGeneration(c *gc.C) {
+	ctrl := gomock.NewController(c)
+	defer ctrl.Finish()
+
+	resultSource := params.ErrorResult{}
+	arg := params.Entity{Tag: s.tag.String()}
+
+	caller := mocks.NewMockAPICallCloser(ctrl)
```
Comment (manadart) [inline api/modelgeneration/modelgeneration_test.go]: The most elegant pattern I've found for these repeated blocks of controller set-up and base assertions is to make the mock(s) a member of the suite and use a method like:
```go
func (s *modelGenerationSuite) setUpMocks(c *gc.C) {
    ctrl := gomock.NewController(c)
    
    s.Caller := mocks.NewMockAPICallCloser(ctrl)
    cExp := s.Caller.EXPECT()
    cExp.BestFacadeVersion(gomock.Any()).Return(0).AnyTimes()

    s.FacadeCaller := mocks.NewMockFacadeCaller(ctrl)
    fExp := s.FacadeCaller.EXPECT()
    fExp.RawAPICaller().Return(caller).AnyTimes()
    
    return ctrl
}
```
Then add the top of the tests, use:
```go
    defer s.setupMocks(c).Finish()
```

File: apiserver/facades/client/modelgeneration/interface.go [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+)
+
+// ModelGeneration defines the methods exported by the model generation API facade.
+type ModelGeneration interface {
+	AddGeneration() (params.ErrorResult, error)
+	AdvanceGeneration(args params.Entities) (params.ErrorResults, error)
+	SwitchGeneration(arg params.GenerationVersionArg) (params.ErrorResult, error)
+}
+
+// ModelGenerationState represents the state of an model required by the ModelGeneration.
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/interface.go]: You could have one generation statement with all of the desired interfaces going into a single file, but it's all the same.

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+		//resources: resources,
+		authorizer: authorizer,
+		//check:          common.NewBlockChecker(st),
+		isControllerAdmin: isAdmin,
+		apiUser:           apiUser,
+		model:             m,
+	}, nil
+}
+
+// authCheck checks if the user is acting on their own behalf, or if they
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: Is this dead, or waiting for some feedback regarding which users should be authorised?

File: apiserver/facades/client/modelgeneration/modelgeneration.go [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+			return results, errors.Trace(err)
+		}
+	}
+
+	ok, err := generation.CanMakeCurrent()
+	if err != nil {
+		return results, errors.Trace(err)
+	}
+	if ok {
+		return results, generation.MakeCurrent()
```
Comment (manadart) [inline apiserver/facades/client/modelgeneration/modelgeneration.go]: I guess we will add the config copy logic to `MakeCurrent` as we progress...

File: cmd/juju/model/addgeneration_test.go [hunk 5]
```diff
@@ -80,22 +81,19 @@ func (s *AddGenerationSuite) TestRunCommand(c *gc.C) {
...
 }
 
 func (s *AddGenerationSuite) TestRunCommandFail(c *gc.C) {
-	c.Skip("Until apiserver ModelGeneration.AddGeneration() implemented")
-
 	mockController, mockAddGenerationCommandAPI := setUpMocks(c)
 	defer mockController.Finish()
 
-	mockAddGenerationCommandAPI.EXPECT().AddGeneration().Return(errors.Errorf("failme")).Times(1)
+	mockAddGenerationCommandAPI.EXPECT().AddGeneration(gomock.Any()).Return(errors.Errorf("failme")).Times(1)
```
Comment (manadart) [inline cmd/juju/model/addgeneration_test.go]: Implicit times(1).

File: cmd/juju/model/switchgeneration.go [hunk 6]
```diff
@@ -104,12 +105,23 @@ func (c *switchGenerationCommand) Run(ctx *cmd.Context) error {
...
+		return err
+	}
 
-	msg := fmt.Sprintf("changes dropped and target generation set to %s\n", c.generation)
+	var msg string
+	switch c.generation {
+	case "next":
+		msg = fmt.Sprintf("target generation set to %s\n", c.generation)
+	case "current":
+		msg = fmt.Sprintf("changes dropped and target generation set to %s\n", c.generation)
```
Comment (manadart) [inline cmd/juju/model/switchgeneration.go]: Is "dropped" the right verbiage to use? Might give some operators a moment of panic.

File: apiserver/facades/client/modelgeneration/modelgeneration.go (hunk 3 above)
Comment (hmlanigan) [reply apiserver/facades/client/modelgeneration/modelgeneration.go]: It's dead, removed.  We should have a  discussion about what the user permissions for this should be.  Right now admin access is required.  I've added a card to Trello for the topic.

File: cmd/juju/model/switchgeneration.go (hunk 6 above)
Comment (hmlanigan) [reply cmd/juju/model/switchgeneration.go]: Must be copy paste error from cancel generation.

File: cmd/juju/model/addgeneration_test.go (hunk 5 above)
Comment (hmlanigan) [reply cmd/juju/model/addgeneration_test.go]: removed.  found the other changes in generation cmd test code too.

File: api/modelgeneration/modelgeneration_test.go (hunk 1 above)
Comment (hmlanigan) [reply api/modelgeneration/modelgeneration_test.go]: nice. thank you.

**Comment by hmlanigan** [comment]
$$merge$$

**Pull request merged**